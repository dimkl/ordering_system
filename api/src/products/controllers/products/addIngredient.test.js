/**
 * @integration-test true
 * @data-factory true
 */

const uuid = require('uuid');

describe("POST /products/ingredients", () => {
  let knex;
  beforeAll(() => knex = require('../../../shared/setupModels')());
  beforeEach(() => knex.raw('truncate products, ingredients cascade;'));
  afterAll(() => knex.destroy());

  it("create product ingredient", async () => {
    const ingredient = await DataFactory.createIngredient();
    const product = await DataFactory.createProduct();

    const response = await request.post(`/products/ingredients`)
      .send({
        ingredient_id: ingredient.uuid,
        product_id: product.uuid,
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchObject({
      id: expect.any(Number),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      uuid: product.uuid,
      title: product.title,
      description: product.description,
      ingredients: expect.arrayContaining([
        expect.objectContaining({
          uuid: ingredient.uuid,
          created_at: expect.any(String),
          updated_at: expect.any(String),
          suitable_for_diet: "all",
          allergen: false
        })
      ]),
    });
    expect(response.body.ingredients.length).toBe(1);
  });

  it("throws 404 for not existing ingredient_id", async () => {
    const product = await DataFactory.createProduct();
    const response = await request.post(`/products/ingredients`)
      .send({
        ingredient_id: uuid.v4(),
        product_id: product.uuid,
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(404);
  });

  it("throws 404 for not existing product_id", async () => {
    const ingredient = await DataFactory.createIngredient();
    const response = await request.post(`/products/ingredients`)
      .send({
        ingredient_id: ingredient.uuid,
        product_id: uuid.v4(),
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(404);
  });

  it("throws validation error with additional properties", async () => {
    const ingredient = await DataFactory.createIngredient();
    const product = await DataFactory.createProduct();

    const response = await request.post(`/products/ingredients`)
      .send({
        ingredient_id: ingredient.uuid,
        product_id: product.uuid,
        created_at: '2022-07-26T22:01:22.539Z'
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });
});