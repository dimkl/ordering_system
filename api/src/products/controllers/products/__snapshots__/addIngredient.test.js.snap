// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /products/ingredients throws validation error with additional properties 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "additionalProperties",
    "message": "must NOT have additional properties",
    "params": Object {
      "additionalProperty": "created_at",
    },
    "schemaPath": "#/additionalProperties",
  },
]
`;
